packages/licenses/src/lockfileToLicenseNodeTree.ts

```diff
@@ -0,0 +1,166 @@
+
+    const packageInstallable = packageIsInstallable(pkgSnapshot.id ?? depPath, {
+      name,
+      version,
+      cpu: pkgSnapshot.cpu,
+      os: pkgSnapshot.os,
+      libc: pkgSnapshot.libc,
+    }, {
+      optional: pkgSnapshot.optional ?? false,
+      lockfileDir: options.dir,
+    })
+
+    // If the package is not installable on the given platform, we ignore the
+    // package, typically the case for platform prebuild packages
+    if (!packageInstallable) {
+      continue
+    }
+
+    const packageInfo = await options.getPackageInfo(
+      {
+        name,
+        version,
+        depPath,
+        snapshot: pkgSnapshot,
+        registries: options.registries,
+      },
+      {
+        storeDir: options.storeDir,
+        virtualStoreDir: options.virtualStoreDir,
+        dir: options.dir,
+        modulesDir: options.modulesDir ?? 'node_modules',
+      }
+    )
+
+    const subdeps = await lockfileToLicenseNode(next(), options)
+
+    const dep: LicenseNode = {
+      name,
+      dev: pkgSnapshot.dev === true,
+      integrity: pkgSnapshot.resolution['integrity'],
+      version,
+      license: packageInfo.license,
+      licenseContents: packageInfo.licenseContents,
+      vendorName: packageInfo.vendorName,
+      vendorUrl: packageInfo.vendorUrl,
+      repository: packageInfo.vendorRepository,
+      dir: packageInfo.path as string,
+    }
+
+    if (Object.keys(subdeps).length > 0) {
+      dep.dependencies = subdeps
+      dep.requires = toRequires(subdeps)
+    }
+
+    // If the package details could be fetched, we consider it part of the tree
+    dependencies[name] = dep
+  }
+
+  return dependencies
+}
+
+/**
+ * Reads the lockfile and converts it in a node tree of information necessary
+ * to generate the licenses summary
+ * @param lockfile the lockfile to process
+ * @param opts     parsing instructions
+ * @returns
+ */
+export async function lockfileToLicenseNodeTree (
+  lockfile: Lockfile,
+  opts: {
+    getPackageInfo: GetPackageInfoFunction
+    include?: { [dependenciesField in DependenciesField]: boolean }
+  } & LicenseExtractOptions
+): Promise<LicenseNodeTree> {
+  const importerWalkers = lockfileWalkerGroupImporterSteps(
+    lockfile,
+    Object.keys(lockfile.importers),
+    { include: opts?.include }
+  )
+  const dependencies = {}
+
+  for (const importerWalker of importerWalkers) {
+    const importerDeps = await lockfileToLicenseNode(importerWalker.step, {
+      storeDir: opts.storeDir,
+      virtualStoreDir: opts.virtualStoreDir,
+      modulesDir: opts.modulesDir,
+      dir: opts.dir,
+      registries: opts.registries,
+      getPackageInfo: opts.getPackageInfo,
+    })
+
+    const depName = importerWalker.importerId
+    dependencies[depName] = {
+      dependencies: importerDeps,
+      requires: toRequires(importerDeps),
+      version: '0.0.0',
+    }
+  }
+
+  const licenseNodeTree: LicenseNodeTree = {
```

Comment:
Why do you create a tree of license nodes if in the output you just list all the packages in a flat list?

Reply:
I will  have a look at changing this

Reply:
Would this be a blocker for the PR to be merged?

Reply:
Isn't that simpler code?